source/material_model/visco_plastic.cc

```diff
@@ -861,23 +955,67 @@ namespace aspect
 
           // Strain weakening parameters
           use_strain_weakening             = prm.get_bool ("Use strain weakening");
-          if (use_strain_weakening)
-            AssertThrow(this->n_compositional_fields() >= 1,
-                        ExcMessage("There must be at least one compositional field. "));
+
+          use_plastic_strain_weakening     = prm.get_bool ("Use plastic strain weakening");
+          if (use_plastic_strain_weakening)
+          {
+            AssertThrow(use_strain_weakening,
+                        ExcMessage("If plastic strain weakening is to be used, strain weakening should also be set to true. "));
+              AssertThrow(this->introspection().compositional_name_exists("plastic_strain") || this->introspection().compositional_name_exists("strain") ||
+                          this->introspection().compositional_name_exists("eii"),
+                          ExcMessage("Material model visco_plastic with both plastic and viscous strain weakening only works if there is a "
+                                     "compositional field called plastic_strain or strain."));
+          }
+
+          use_viscous_strain_weakening     = prm.get_bool ("Use viscous strain weakening");
+          if (use_viscous_strain_weakening)
+          {
+            AssertThrow(use_strain_weakening,
+                        ExcMessage("If viscous strain weakening is to be used, strain weakening should also be set to true. "));
+              AssertThrow(this->introspection().compositional_name_exists("viscous_strain") || this->introspection().compositional_name_exists("strain") ||
+                          this->introspection().compositional_name_exists("eii"),
+                          ExcMessage("Material model visco_plastic with viscous strain weakening only works if there is a "
+                                     "compositional field called viscous_strain or strain."));
+          }
```

Comment:
Nice! This commit came through just as I made a comment on the TODO in `visco_plastic.h`. As the `composition_mask` function make assumptions about the first 1 (total strain) or 2 compositional fields (plastic and viscous strain), should we also enforce that they are list in the correct order here? Alternatively, we could use the names to correctly mark them within `composition_mask`.  

If we do not modify  the current structure`composition_mask`, then I think we should enforce the listed order of the strain fields in the `AssertThrow`.